extensions/wikia/ArticleComments/classes/ArticleComment.class.php

```diff
@@ -197,84 +192,124 @@ static public function newFromId( $id ) {
+		if ( empty( $this->mFirstRevId ) ) {
+			$this->mFirstRevId = $this->getFirstRevID( DB_MASTER );
+		}
 
-			// get revision objects
-			if ( $this->mFirstRevId ) {
-				$this->mFirstRevision = Revision::newFromId( $this->mFirstRevId );
-				if ( !empty( $this->mFirstRevision ) && is_object( $this->mFirstRevision ) && ( $this->mFirstRevision instanceof Revision ) ) { // fix for FB:15198
-					if ( $this->mLastRevId == $this->mFirstRevId ) {
-						// save one db query by just setting them to the same revision object
-						$this->mLastRevision = $this->mFirstRevision;
-					} else {
-						$this->mLastRevision = Revision::newFromId( $this->mLastRevId );
-						if ( empty( $this->mLastRevision ) || !is_object( $this->mLastRevision ) || !( $this->mLastRevision instanceof Revision ) ) {
-							$result = false;
-						}
-					}
-				} else {
-					$result = false;
-				}
-			} else {
-				$result = false;
-			}
+		return !empty( $this->mFirstRevId );
+	}
 
-			// get user that created this comment
-			if ( $this->mFirstRevision ) {
-				$this->mUser = User::newFromId( $this->mFirstRevision->getUser() );
-				$this->mUser->setName( $this->mFirstRevision->getUserText() );
-			} else {
-				$result = false;
-			}
+	private function loadLastRevId( $useMaster = false ) {
+		if ( empty( $this->mTitle ) ) {
+			return false;
+		}
 
-			if ( empty( $this->mFirstRevision ) || empty( $this->mLastRevision ) ) {
-				wfProfileOut( __METHOD__ );
-				return false;
-			}
+		// Exit early if we've already loaded this
+		if ( !empty( $this->mLastRevId ) ) {
+			return true;
+		}
 
-			$rawText = $this->mLastRevision->getText();
-			$this->parseText( $rawText );
-		} else { // null title
-			$result = false;
+		if ( !$useMaster ) {
+			$this->mLastRevId = $this->mTitle->getLatestRevID();
 		}
 
-		wfProfileOut( __METHOD__ );
-		return $result;
+		// Fall back to master if not on slave or if we wanted master in the first place
+		if ( empty( $this->mLastRevId ) ) {
+			$this->mLastRevId = $this->mTitle->getLatestRevID( Title::GAID_FOR_UPDATE );
+		}
+
+		// Use first rev ID as last resort
+		if ( empty( $this->mLastRevId ) ) {
+			$this->loadFirstRevId();
+			$this->mLastRevId = $this->mFirstRevId;
+		}
+
+		return !empty( $this->mLastRevId );
+	}
+
+	private function loadFirstRevision() {
+		if ( !empty( $this->mFirstRevision ) ) {
+			return true;
+		}
+
+		// get revision objects
+		$this->mFirstRevision = Revision::newFromId( $this->mFirstRevId );
+
+		return !empty( $this->mFirstRevision ) && $this->mFirstRevision instanceof Revision;
+	}
+
+	private function loadLastRevision() {
+		if ( !empty( $this->mLastRevision ) ) {
+			return true;
+		}
+
+		if ( $this->mLastRevId == $this->mFirstRevId ) {
+			// save one db query by just setting them to the same revision object
+			$this->mLastRevision = $this->mFirstRevision;
+		} else {
+			$this->mLastRevision = Revision::newFromId( $this->mLastRevId );
```

Comment:
Ok nope.  Other classes are way too involved in the inner workings of this class to make these getters work the way they should without doing a MASSIVE re-architecture.  Several classes currently expect to call ```load``` directly and then do some error checking if it fails.

This means that we can't lazy load `firstRevID` here and `lastRevID` there as needed because there is one place where all this code expects failure to happen, and its not when requesting these values.

Why not load everything when a constructor is called and fail to return an `ArticleComment` object?  Well for one, the `load` method takes a `useMaster` flag that determines how the data is loaded (e.g. master or slave) and we'd have to add that to every constructor as a second, awkward argument.

Also, not every comment gets or needs to be loaded currently.  Lazy loading would fix this but again, current callers are not equipped to handle errors outside `load`.

Also, in one place (`ArticleCommentList`) the code pre-loads a bunch of firstRevIds values from the DB and sets the value directly on an array of `ArticleComment` objects.  This improves performance by not doing several separate lookups but it completely ruins any abstraction.  This functionality would be better as a constructor in `ArticleComment` but it happens deep within some function calls using some nasty `for` loops.

My current updates (with `load*` methods) don't change how the code works.  I basically just took the code from one `load` method and made several smaller `load` methods.  I think this is the most we can change with out having to basically rewrite everything.